lib/http/2/compressor.rb

```diff
@@ -545,22 +559,38 @@ def header(buf)
         header
       end
 
+      FORBIDDEN_HEADERS = %w[connection te].freeze
+
       # Decodes and processes header commands within provided buffer.
       #
       # @param buf [Buffer]
-      # @return [Array] +[[name, value], ...]+
-      def decode(buf)
+      # @param frame [HTTP2::Frame, nil]
+      # @return [Array] +[[name, value], ...]
+      def decode(buf, frame = nil)
         list = []
         decoding_pseudo_headers = true
-        until buf.empty?
-          next_header = @cc.process(header(buf))
-          next if next_header.nil?
-          is_pseudo_header = next_header.first.start_with? ':'
-          if !decoding_pseudo_headers && is_pseudo_header
-            fail ProtocolError, 'one or more pseudo headers encountered after regular headers'
+        @cc.listen_on_table do
+          until buf.empty?
+            field, value = @cc.process(header(buf))
+            next if field.nil?
+            is_pseudo_header = field.start_with? ':'
+            if !decoding_pseudo_headers && is_pseudo_header
+              fail ProtocolError, 'one or more pseudo headers encountered after regular headers'
+            end
+            decoding_pseudo_headers = is_pseudo_header
+            if FORBIDDEN_HEADERS.include?(field)
+              fail ProtocolError, "invalid header received: #{field}"
+            end
+            case field
+            when ':method'
+              frame[:method] = value
+            when 'content-length'
+              frame[:content_length] = Integer(value)
```

Comment:
this was done as an optimization, as method, content length and trailer header are all used in the stream for other checks, and decode only returns an array of arrays, which is not the best for lookups.

Reply:
this is not treating content length as a pseudo-header; it is rather declaring it explicitly in the frame, so that it can be fetched in an easy way to perform calculations in the stream. If I don't do this, content length will be in the headers array, and I'll have to traverse it to reach it. As stated, it was an optimization for this particular use case (as where the cases of method and trailer).